R/cohensFleissKappa.R

```diff
@@ -0,0 +1,184 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+cohensFleissKappa <- function(jaspResults, dataset, options) {
+  
+  ready <- length(options[["variables"]]) > 1
+  
+  dataset <- .readDataCohensFleissKappa(dataset, options)
+    
+  if (options[["cohensKappa"]])
+    jaspResults[["cohensKappa"]] <- .computeCohensKappaTable(dataset, options, ready)
+  
+  if (options[["fleissKappa"]])
+    jaspResults[["fleissKappa"]] <- .computeFleissKappaTable(dataset, options, ready)
+  
+  return()
+}
+
+# Read in the dataset (copied from .reliabilityReadData)
+.readDataCohensFleissKappa <- function(dataset, options) {
+  variables <- unlist(options[["variables"]])
+  if (is.null(dataset)) {
+    dataset <- .readDataSetToEnd(columns.as.factor = variables)
+  }
+  return(dataset)
+}
+
+.computeCohensKappaTable <- function(dataset, options, ready) {
+  
+  weighted <- options[["cohensWeightedOrNot"]] == "cohensWeighted"
+  
+  weightedString <- ifelse(weighted, "Weighted", "Unweighted")
+  
+  # Create the JASP Table
+  jaspTable <- createJaspTable(title = gettextf("Cohen's %s Kappa", casefold(weightedString)))
+  jaspTable$addColumnInfo(name = "ratings", title = gettext("Ratings"), type = "string")
+  jaspTable$addColumnInfo(name = "cKappa", title = gettextf("%s Kappa", weightedString), type = "number")
+  jaspTable$position <- 1
+  
+  #dependencies
+  jaspTable$dependOn(
+    options = c(
+      "variables",
+      "cohensKappa",
+      "cohensWeightedOrNot",
+      "kappaIntervalOn",
+      "kappaConfidenceIntervalValue"
+    )
+  )
+  
+  
+  formattedCIPercent <- format(
+    100 * options[["kappaConfidenceIntervalValue"]],
+    digits = 3,
+    drop0trailing = TRUE
+  )
+  
+  
+  if (ready) {
+    #calculate Cohen's Kappas
+    nPairs <- ncol(dataset) * (ncol(dataset) - 1) / 2
+    out_kappa <- psych::cohen.kappa(dataset, alpha = 1 - options[["kappaConfidenceIntervalValue"]])
+    if (nPairs == 1) {
+      allKappaData <- list(out_kappa)
+      allPairStrings <- paste(options[["variables"]], collapse = " - ")
+    } else {
+      allKappaData <- out_kappa[2:(nPairs + 1)]
+      allPairStrings <- sub(" ", " - ", names(out_kappa[2:(nPairs + 1)]))
+    }
+
+    #Extract Kappas and CIs
+    allKappas <- c()
+    allLowerBounds <- c()
+    allUpperBounds <- c()
+    
+    for (i in allKappaData) {
+      kappaData <- i$confid
+      k <- ifelse(weighted, 2, 1)
+      allKappas <- c(allKappas, kappaData[k, 2])
+      allLowerBounds <- c(allLowerBounds, kappaData[k, 1])
+      allUpperBounds <- c(allUpperBounds, kappaData[k, 3])
+      averageKappa <- ifelse(weighted, out_kappa$av.wt, out_kappa$av.kappa)
```

Comment:
When I do an analysis with only two variables I get an error traced back to this line.

![Screenshot 2021-12-13 at 11 02 31](https://user-images.githubusercontent.com/38500953/145791823-7ad0bf44-bdbe-4c65-a803-990ac91d7916.png)